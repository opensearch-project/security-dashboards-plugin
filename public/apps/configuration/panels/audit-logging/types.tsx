/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export interface GeneralSettings {
  // layer
  enable_rest?: boolean;
  enable_transport?: boolean;
  disabled_rest_categories?: string[];
  disabled_transport_categories?: string[];

  // attribute
  resolve_bulk_requests?: boolean;
  resolve_indices?: boolean;
  log_request_body?: boolean;
  exclude_sensitive_headers?: boolean;

  // ignore
  ignore_users?: string[];
  ignore_requests?: string[];
}

export interface ComplianceSettings {
  enabled?: boolean;

  // read
  read_metadata_only?: boolean;
  read_ignore_users?: string[];
  read_watched_fields?: {
    [key: string]: string;
  };

  // write
  write_metadata_only?: boolean;
  write_log_diffs?: boolean;
  write_ignore_users?: string[];
  write_watched_indices?: string[];
}

export interface AuditLoggingSettings {
  enabled?: boolean;
  audit?: GeneralSettings;
  compliance?: ComplianceSettings;
}
