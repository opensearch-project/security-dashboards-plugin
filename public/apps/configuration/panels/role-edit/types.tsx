/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiComboBoxOptionOption } from '@elastic/eui';

export type ComboBoxOptions = EuiComboBoxOptionOption[];
export type FieldLevelSecurityMethod = 'exclude' | 'include';
export interface RoleIndexPermissionStateClass {
  indexPatterns: ComboBoxOptions;
  docLevelSecurity: string;
  fieldLevelSecurityMethod: FieldLevelSecurityMethod;
  fieldLevelSecurityFields: ComboBoxOptions;
  maskedFields: ComboBoxOptions;
  allowedActions: ComboBoxOptions;
}

export enum TenantPermissionType {
  None = '',
  Read = 'r',
  Write = 'w',
  Full = 'rw',
}

export interface RoleTenantPermissionStateClass {
  tenantPatterns: ComboBoxOptions;
  permissionType: TenantPermissionType;
}
