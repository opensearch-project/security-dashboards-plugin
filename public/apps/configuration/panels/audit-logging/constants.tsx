/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */
import { pullAll } from 'lodash';

export const CONFIG_LABELS = {
  AUDIT_LOGGING: 'Audit logging',
  GENERAL_SETTINGS: 'General settings',
  LAYER_SETTINGS: 'Layer settings',
  ATTRIBUTE_SETTINGS: 'Attribute settings',
  IGNORE_SETTINGS: 'Ignore settings',
  COMPLIANCE_SETTINGS: 'Compliance settings',
  COMPLIANCE_MODE: 'Compliance mode',
  COMPLIANCE_CONFIG_SETTINGS: 'Config',
  COMPLIANCE_READ: 'Read',
  COMPLIANCE_WRITE: 'Write',
};

export const RESPONSE_MESSAGES = {
  FETCH_ERROR_TITLE: 'Sorry, there was an error fetching audit configuration.',
  FETCH_ERROR_MESSAGE:
    'Please ensure hot reloading of audit configuration is enabled in the security plugin.',
  UPDATE_SUCCESS: 'Audit configuration was successfully updated.',
  UPDATE_FAILURE: 'Audit configuration could not be updated. Please check configuration.',
};

export interface SettingMapItem {
  key: string;
  value: string[];
}

export interface SettingContent {
  title: string;
  path: string;
  description: string;
  type: string;
  options?: string[];
  placeHolder?: string;
  code?: string;
  error?: string;
}

const REST_LAYER: SettingContent = {
  title: 'REST layer',
  path: 'audit.enable_rest',
  description: 'Enable or disable auditing events that happen on the REST layer.',
  type: 'bool',
};

const REST_DISABLED_CATEGORIES: SettingContent = {
  title: 'REST disabled categories',
  path: 'audit.disabled_rest_categories',
  description:
    'Specify audit categories which must be ignored on the REST layer. Modifying these could result in ' +
    'significant overhead.',
  type: 'array',
  options: [
    'BAD_HEADERS',
    'FAILED_LOGIN',
    'MISSING_PRIVILEGES',
    'GRANTED_PRIVILEGES',
    'SSL_EXCEPTION',
    'AUTHENTICATED',
  ],
  placeHolder: 'Select categories',
};

export function excludeFromDisabledRestCategories(optionsToExclude: string[]) {
  pullAll(REST_DISABLED_CATEGORIES.options || [], optionsToExclude);
}

const TRANSPORT_LAYER: SettingContent = {
  title: 'Transport layer',
  path: 'audit.enable_transport',
  description: 'Enable or disable auditing events that happen on the transport layer.',
  type: 'bool',
};

const TRANSPORT_DISABLED_CATEGORIES: SettingContent = {
  title: 'Transport disabled categories',
  path: 'audit.disabled_transport_categories',
  description:
    'Specify audit categories which must be ignored on the transport layer. Modifying these could result ' +
    'in significant overhead.',
  type: 'array',
  options: [
    'BAD_HEADERS',
    'FAILED_LOGIN',
    'GRANTED_PRIVILEGES',
    'INDEX_EVENT',
    'MISSING_PRIVILEGES',
    'SSL_EXCEPTION',
    'OPENDISTRO_SECURITY_INDEX_ATTEMPT',
    'AUTHENTICATED',
  ],
  placeHolder: 'Select categories',
};

export function excludeFromDisabledTransportCategories(optionsToExclude: string[]) {
  pullAll(TRANSPORT_DISABLED_CATEGORIES.options || [], optionsToExclude);
}

const BULK_REQUESTS: SettingContent = {
  title: 'Bulk requests',
  path: 'audit.resolve_bulk_requests',
  description:
    'Resolve bulk requests during auditing of requests. Enabling this will generate a log for each ' +
    'document request which could result in significant overhead.',
  type: 'bool',
};

const REQUEST_BODY: SettingContent = {
  title: 'Request body',
  path: 'audit.log_request_body',
  description: 'Include request body during auditing of requests.',
  type: 'bool',
};

const RESOLVE_INDICES: SettingContent = {
  title: 'Resolve indices',
  path: 'audit.resolve_indices',
  description: 'Resolve indices during auditing of requests.',
  type: 'bool',
};

const SENSITIVE_HEADERS: SettingContent = {
  title: 'Sensitive headers',
  path: 'audit.exclude_sensitive_headers',
  description: 'Exclude sensitive headers during auditing. (e.g. authorization header)',
  type: 'bool',
};

const IGNORED_USERS: SettingContent = {
  title: 'Ignored users',
  path: 'audit.ignore_users',
  description:
    'Users to ignore during auditing. Changing the defaults could result in significant overhead.',
  type: 'array',
  placeHolder: 'Add users or user patterns',
};

const IGNORED_REQUESTS: SettingContent = {
  title: 'Ignored requests',
  path: 'audit.ignore_requests',
  description: 'Request patterns to ignore during auditing.',
  type: 'array',
  placeHolder: 'Add request patterns',
};

const ENABLED: SettingContent = {
  title: 'Compliance logging',
  path: 'compliance.enabled',
  description: 'Enable or disable compliance logging.',
  type: 'bool',
};

const INTERNAL_CONFIG: SettingContent = {
  title: 'Internal config logging',
  path: 'compliance.internal_config',
  description: 'Enable or disable logging of events on internal security index.',
  type: 'bool',
};

const EXTERNAL_CONFIG: SettingContent = {
  title: 'External config logging',
  path: 'compliance.external_config',
  description: 'Enable or disable logging of external configuration.',
  type: 'bool',
};

const READ_METADATA_ONLY: SettingContent = {
  title: 'Read metadata',
  path: 'compliance.read_metadata_only',
  description: 'Do not log any document fields. Log only metadata of the document.',
  type: 'bool',
};

const READ_IGNORED_USERS: SettingContent = {
  title: 'Ignored users',
  path: 'compliance.read_ignore_users',
  description: 'Users to ignore during auditing.',
  type: 'array',
  placeHolder: 'Add users or user patterns',
};

const READ_WATCHED_FIELDS: SettingContent = {
  title: 'Watched fields',
  path: 'compliance.read_watched_fields',
  description:
    'List the indices and fields to watch during read events. Adding watched fields will generate one log per document' +
    ' access and could result in significant overhead. Sample data content:',
  type: 'map',
  code: `{
  "index-name-pattern": ["field-name-pattern"],
  "logs*": ["message"],
  "twitter": ["id", "user*"]
}`,
  error: 'Invalid content. Please check sample data content.',
};

const WRITE_METADATA_ONLY: SettingContent = {
  title: 'Write metadata',
  path: 'compliance.write_metadata_only',
  description: 'Do not log any document content. Log only metadata of the document.',
  type: 'bool',
};

const WRITE_LOG_DIFFS: SettingContent = {
  title: 'Log diffs',
  path: 'compliance.write_log_diffs',
  description: 'Log diffs for document updates.',
  type: 'bool',
};

const WRITE_IGNORED_USERS: SettingContent = {
  title: 'Ignored users',
  path: 'compliance.write_ignore_users',
  description: 'Users to ignore during auditing.',
  type: 'array',
  placeHolder: 'Add users or user patterns',
};

const WRITE_WATCHED_FIELDS: SettingContent = {
  title: 'Watch indices',
  path: 'compliance.write_watched_indices',
  description:
    'List the indices to watch during write events. Adding watched indices will generate one log per ' +
    'document access and could result in significant overhead.',
  type: 'array',
  placeHolder: 'Add indices',
};

export const CONFIG = {
  ENABLED: {
    title: 'Enable audit logging',
    path: 'enabled',
    description: 'Enable or disable audit logging',
    type: 'bool',
  },
  AUDIT: {
    REST_LAYER,
    REST_DISABLED_CATEGORIES,
    TRANSPORT_LAYER,
    TRANSPORT_DISABLED_CATEGORIES,
    BULK_REQUESTS,
    REQUEST_BODY,
    RESOLVE_INDICES,
    SENSITIVE_HEADERS,
    IGNORED_USERS,
    IGNORED_REQUESTS,
  },
  COMPLIANCE: {
    ENABLED,
    INTERNAL_CONFIG,
    EXTERNAL_CONFIG,
    READ_METADATA_ONLY,
    READ_IGNORED_USERS,
    READ_WATCHED_FIELDS,
    WRITE_METADATA_ONLY,
    WRITE_LOG_DIFFS,
    WRITE_IGNORED_USERS,
    WRITE_WATCHED_FIELDS,
  },
};

export interface SettingGroup {
  title: string;
  settings: SettingContent[];
}

export const LAYER_SETTINGS: SettingGroup = {
  title: CONFIG_LABELS.LAYER_SETTINGS,
  settings: [
    CONFIG.AUDIT.REST_LAYER,
    CONFIG.AUDIT.REST_DISABLED_CATEGORIES,
    CONFIG.AUDIT.TRANSPORT_LAYER,
    CONFIG.AUDIT.TRANSPORT_DISABLED_CATEGORIES,
  ],
};

export const ATTRIBUTE_SETTINGS: SettingGroup = {
  title: CONFIG_LABELS.ATTRIBUTE_SETTINGS,
  settings: [
    CONFIG.AUDIT.BULK_REQUESTS,
    CONFIG.AUDIT.REQUEST_BODY,
    CONFIG.AUDIT.RESOLVE_INDICES,
    CONFIG.AUDIT.SENSITIVE_HEADERS,
  ],
};

export const IGNORE_SETTINGS: SettingGroup = {
  title: CONFIG_LABELS.IGNORE_SETTINGS,
  settings: [CONFIG.AUDIT.IGNORED_USERS, CONFIG.AUDIT.IGNORED_REQUESTS],
};

export const COMPLIANCE_CONFIG_MODE_SETTINGS: SettingGroup = {
  title: CONFIG_LABELS.COMPLIANCE_MODE,
  settings: [CONFIG.COMPLIANCE.ENABLED],
};

export const COMPLIANCE_CONFIG_SETTINGS: SettingGroup = {
  title: CONFIG_LABELS.COMPLIANCE_CONFIG_SETTINGS,
  settings: [CONFIG.COMPLIANCE.INTERNAL_CONFIG, CONFIG.COMPLIANCE.EXTERNAL_CONFIG],
};

export const COMPLIANCE_SETTINGS_READ: SettingGroup = {
  title: CONFIG_LABELS.COMPLIANCE_READ,
  settings: [
    CONFIG.COMPLIANCE.READ_METADATA_ONLY,
    CONFIG.COMPLIANCE.READ_IGNORED_USERS,
    CONFIG.COMPLIANCE.READ_WATCHED_FIELDS,
  ],
};

export const COMPLIANCE_SETTINGS_WRITE: SettingGroup = {
  title: CONFIG_LABELS.COMPLIANCE_WRITE,
  settings: [
    CONFIG.COMPLIANCE.WRITE_METADATA_ONLY,
    CONFIG.COMPLIANCE.WRITE_LOG_DIFFS,
    CONFIG.COMPLIANCE.WRITE_IGNORED_USERS,
    CONFIG.COMPLIANCE.WRITE_WATCHED_FIELDS,
  ],
};

export const SETTING_GROUPS = {
  AUDIT_SETTINGS: {
    settings: [CONFIG.ENABLED],
  },
  LAYER_SETTINGS,
  ATTRIBUTE_SETTINGS,
  IGNORE_SETTINGS,
  COMPLIANCE_CONFIG_MODE_SETTINGS,
  COMPLIANCE_CONFIG_SETTINGS,
  COMPLIANCE_SETTINGS_READ,
  COMPLIANCE_SETTINGS_WRITE,
};

export const SUB_URL_FOR_GENERAL_SETTINGS_EDIT = '/edit/generalSettings';
export const SUB_URL_FOR_COMPLIANCE_SETTINGS_EDIT = '/edit/complianceSettings';
