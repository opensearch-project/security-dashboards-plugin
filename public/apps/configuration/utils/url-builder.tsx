/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { Action } from '../types';
import { ResourceType } from '../../../../common';

/**
 * Build hash based url, encode the resourceId part
 * e.g.
 *   buildUrl() => "/" (landing page)
 *   buildUrl(ResourceType.roles) => "/roles" (role listing page)
 *   buildUrl(ResourceType.roles, Action.create) => "/roles/create" (role creation page)
 *   buildUrl(ResourceType.roles, Action.view, "../someRole") => "/roles/view/..%2FsomeRole" (role detail page)
 *   buildUrl(ResourceType.roles, Action.view, "someRole", SubAction.mapuser) => "/roles/view/someRole/mapuser" (Map User page)
 * edge case (wrong usage) fallbacks:
 *   buildUrl(undefined, Action.create) => ""
 *   buildUrl(ResourceType.roles, undefined, "someRole") => "/roles"
 */
export function buildUrl(
  resouceType?: ResourceType,
  action?: Action,
  resourceId?: string,
  subAction?: string
) {
  const encodedResouceId = resourceId ? encodeURIComponent(resourceId) : undefined;
  const rawContents = [resouceType, action, encodedResouceId, subAction];
  const contents = [];
  for (const content of rawContents) {
    if (content === undefined) {
      break;
    }
    contents.push(content);
  }
  return '/' + contents.join('/');
}

export function buildHashUrl(
  resouceType?: ResourceType,
  action?: Action,
  resourceId?: string,
  subAction?: string
) {
  return '#' + buildUrl(resouceType, action, resourceId, subAction);
}
