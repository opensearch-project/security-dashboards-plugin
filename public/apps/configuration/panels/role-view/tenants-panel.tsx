/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React, { useState, useEffect } from 'react';
import {
  EuiInMemoryTable,
  EuiLink,
  EuiText,
  EuiGlobalToastList,
  EuiBasicTableColumn,
  EuiEmptyPrompt,
  EuiButton,
} from '@elastic/eui';
import { CoreStart } from 'opensearch-dashboards/public';
import { getCurrentUser } from '../../../../utils/auth-info-utils';
import { PanelWithHeader } from '../../utils/panel-with-header';
import {
  RoleTenantPermissionView,
  RoleTenantPermissionDetail,
  ResourceType,
  Action,
} from '../../types';
import { truncatedListView, tableItemsUIProps } from '../../utils/display-utils';
import {
  fetchTenants,
  transformTenantData,
  transformRoleTenantPermissionData,
  selectTenant,
} from '../../utils/tenant-utils';
import { DocLinks, RoleViewTenantInvalidText } from '../../constants';
import { PageId } from '../../types';
import { getNavLinkById } from '../../../../services/chrome_wrapper';
import { useToastState, createUnknownErrorToast } from '../../utils/toast-utils';
import { showTableStatusMessage } from '../../utils/loading-spinner-utils';
import { buildHashUrl } from '../../utils/url-builder';

interface RoleViewTenantsPanelProps {
  roleName: string;
  tenantPermissions: RoleTenantPermissionView[];
  errorFlag: boolean;
  coreStart: CoreStart;
  loading: boolean;
  isReserved: boolean;
}

export function TenantsPanel(props: RoleViewTenantsPanelProps) {
  const [tenantPermissionDetail, setTenantPermissionDetail] = React.useState<
    RoleTenantPermissionDetail[]
  >([]);
  const [currentUsername, setCurrentUsername] = React.useState('');
  const [toasts, addToast, removeToast] = useToastState();
  const [errorFlag, setErrorFlag] = React.useState(props.errorFlag);

  React.useEffect(() => {
    const fetchData = async () => {
      try {
        const rawTenantData = await fetchTenants(props.coreStart.http);
        const processedTenantData = transformTenantData(rawTenantData);
        setTenantPermissionDetail(
          transformRoleTenantPermissionData(props.tenantPermissions, processedTenantData)
        );
        const currentUser = await getCurrentUser(props.coreStart.http);
        setCurrentUsername(currentUser);
      } catch (e) {
        console.log(e);
        setErrorFlag(true);
      }
    };

    fetchData();
  }, [props]);

  const viewDashboard = async (tenantName: string) => {
    try {
      await selectTenant(props.coreStart.http, {
        tenant: tenantName,
        username: currentUsername,
      });
      window.location.href = getNavLinkById(props.coreStart, PageId.dashboardId);
    } catch (e) {
      console.log(e);
      addToast(createUnknownErrorToast('viewDashboard', `view dashboard for ${tenantName} tenant`));
    }
  };

  const viewVisualization = async (tenantName: string) => {
    try {
      await selectTenant(props.coreStart.http, {
        tenant: tenantName,
        username: currentUsername,
      });
      window.location.href = getNavLinkById(props.coreStart, PageId.visualizationId);
    } catch (e) {
      console.log(e);
      addToast(
        createUnknownErrorToast('viewVisualization', `view visualization for ${tenantName} tenant`)
      );
    }
  };

  const columns: Array<EuiBasicTableColumn<RoleTenantPermissionDetail>> = [
    {
      field: 'tenant_patterns',
      name: 'Name',
      render: truncatedListView(tableItemsUIProps),
      truncateText: true,
      sortable: true,
    },
    {
      field: 'description',
      name: 'Description',
      truncateText: true,
    },
    {
      field: 'permissionType',
      name: 'Read/write permission',
    },
    {
      field: 'tenantValue',
      name: 'Dashboard',
      render: (tenant: string) => {
        if (tenant === RoleViewTenantInvalidText) {
          return (
            <>
              <EuiText size="s">{RoleViewTenantInvalidText}</EuiText>
            </>
          );
        }
        return (
          <>
            <EuiLink data-test-subj="view-dashboard" onClick={() => viewDashboard(tenant)}>
              View dashboard
            </EuiLink>
          </>
        );
      },
    },
    {
      field: 'tenantValue',
      name: 'Visualizations',
      render: (tenant: string) => {
        if (tenant === RoleViewTenantInvalidText) {
          return (
            <>
              <EuiText size="s">{RoleViewTenantInvalidText}</EuiText>
            </>
          );
        }
        return (
          <>
            <EuiLink data-test-subj="view-visualizations" onClick={() => viewVisualization(tenant)}>
              View visualizations
            </EuiLink>
          </>
        );
      },
    },
  ];

  const emptyListMessage = (
    <EuiEmptyPrompt
      title={<h3>No tenant permission</h3>}
      titleSize="s"
      actions={
        <EuiButton
          data-test-subj="addTenantPermission"
          disabled={props.isReserved}
          onClick={() => {
            window.location.href = buildHashUrl(ResourceType.roles, Action.edit, props.roleName);
          }}
        >
          Add tenant permission
        </EuiButton>
      }
    />
  );

  const headerText = 'Tenant permissions';
  return (
    <>
      <PanelWithHeader
        headerText={headerText}
        headerSubText="Tenants in OpenSearch Dashboards are spaces for saving index patterns, visualizations, dashboards, and other OpenSearch Dashboards objects.
        Tenants are useful for safely sharing your work with other OpenSearch Dashboards users. You can control which roles have access to a tenant and whether those roles have read or write access."
        helpLink={DocLinks.TenantPermissionsDoc}
        count={tenantPermissionDetail.length}
      >
        <EuiInMemoryTable
          data-test-subj="tenant-permission-container"
          tableLayout={'auto'}
          loading={tenantPermissionDetail === [] && !errorFlag}
          columns={columns}
          items={tenantPermissionDetail}
          sorting={{ sort: { field: 'type', direction: 'asc' } }}
          error={errorFlag ? 'Load data failed, please check console log for more detail.' : ''}
          message={showTableStatusMessage(props.loading, props.tenantPermissions, emptyListMessage)}
        />
      </PanelWithHeader>
      <EuiGlobalToastList toasts={toasts} toastLifeTimeMs={10000} dismissToast={removeToast} />
    </>
  );
}
