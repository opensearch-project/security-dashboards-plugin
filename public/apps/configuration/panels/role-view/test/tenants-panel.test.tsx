/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { mount, shallow } from 'enzyme';
import { TenantsPanel } from '../tenants-panel';
import { EuiEmptyPrompt, EuiInMemoryTable, EuiTableFieldDataColumnType } from '@elastic/eui';
import { buildHashUrl } from '../../../utils/url-builder';
import { Action, RoleTenantPermissionDetail } from '../../../types';
import { ResourceType } from '../../../../../../common';
import { RoleViewTenantInvalidText } from '../../../constants';

jest.mock('../../../utils/tenant-utils');
jest.mock('../../../../../utils/auth-info-utils');
// eslint-disable-next-line
const mockTenantUtils = require('../../../utils/tenant-utils');
// eslint-disable-next-line
const mockAuthInfoUtils = require('../../../../../utils/auth-info-utils');

describe('Role view - tenant panel', () => {
  const setState = jest.fn();
  const mockCoreStart = {
    http: 1,
  };
  beforeEach(() => {
    jest.spyOn(React, 'useState').mockImplementation((initialValue) => [initialValue, setState]);
  });

  describe('Tenant permission list', () => {
    const sampleRoleName = 'role1';
    it('render empty', () => {
      const wrapper = shallow(
        <TenantsPanel
          roleName={sampleRoleName}
          tenantPermissions={[]}
          errorFlag={false}
          coreStart={mockCoreStart as any}
          loading={false}
          isReserved={false}
        />
      );

      expect(wrapper.find(EuiInMemoryTable).prop('items').length).toBe(0);
    });

    it('it renders a empty prompt when tenantPermissions is empty', () => {
      const wrapper = mount(
        <TenantsPanel
          roleName={sampleRoleName}
          tenantPermissions={[]}
          errorFlag={false}
          coreStart={mockCoreStart as any}
          loading={false}
          isReserved={false}
        />
      );

      const prompt = wrapper
        .find('[data-test-subj="tenant-permission-container"] tbody EuiEmptyPrompt')
        .first()
        .getElement();
      expect(prompt.type).toBe(EuiEmptyPrompt);
    });

    it('Add tenant permission button is disabled for reserved role', () => {
      const wrapper = mount(
        <TenantsPanel
          roleName={sampleRoleName}
          tenantPermissions={[]}
          errorFlag={false}
          coreStart={mockCoreStart as any}
          loading={false}
          isReserved={true}
        />
      );

      const prompt = wrapper
        .find('[data-test-subj="tenant-permission-container"] tbody EuiEmptyPrompt')
        .first()
        .getElement();
      const component = shallow(prompt);
      const button = component.find('[data-test-subj="addTenantPermission"]');
      expect(button.prop('disabled')).toBe(true);
    });

    it('should render to role edit page when click on Add tenant permission button', () => {
      const wrapper = mount(
        <TenantsPanel
          roleName={sampleRoleName}
          tenantPermissions={[]}
          errorFlag={false}
          coreStart={mockCoreStart as any}
          loading={false}
          isReserved={true}
        />
      );

      const prompt = wrapper
        .find('[data-test-subj="tenant-permission-container"] tbody EuiEmptyPrompt')
        .first()
        .getElement();
      const component = shallow(prompt);
      component.find('[data-test-subj="addTenantPermission"]').simulate('click');
      expect(window.location.hash).toBe(
        buildHashUrl(ResourceType.roles, Action.edit, sampleRoleName)
      );
    });

    it('click on View Dashboard link', () => {
      const tenantPermissions = [{ tenant_patterns: ['tenant1'], permissionType: 'dummy' }];
      const wrapper = mount(
        <TenantsPanel
          roleName={sampleRoleName}
          tenantPermissions={tenantPermissions}
          errorFlag={false}
          coreStart={mockCoreStart as any}
          loading={false}
          isReserved={true}
        />
      );

      const prompt = wrapper
        .find('[data-test-subj="tenant-permission-container"]')
        .first()
        .getElement();
      const component = shallow(prompt);
      const columns = component.prop<
        Array<EuiTableFieldDataColumnType<RoleTenantPermissionDetail>>
      >('columns');
      const viewDashboardRenderer = columns[3].render as (tenant: string) => JSX.Element;
      const Container = (props: { tenant: string }) => viewDashboardRenderer(props.tenant);
      const result = shallow(<Container tenant={'tenant1'} />);
      result.find('[data-test-subj="view-dashboard"]').simulate('click');
      expect(mockTenantUtils.selectTenant).toHaveBeenCalled();
    });

    it('click on View Visualizations link', () => {
      const tenantPermissions = [{ tenant_patterns: ['tenant1'], permissionType: 'dummy' }];
      const wrapper = mount(
        <TenantsPanel
          roleName={sampleRoleName}
          tenantPermissions={tenantPermissions}
          errorFlag={false}
          coreStart={mockCoreStart as any}
          loading={false}
          isReserved={true}
        />
      );

      const prompt = wrapper
        .find('[data-test-subj="tenant-permission-container"]')
        .first()
        .getElement();
      const component = shallow(prompt);
      const columns = component.prop<
        Array<EuiTableFieldDataColumnType<RoleTenantPermissionDetail>>
      >('columns');
      const viewVisualizationRenderer = columns[4].render as (tenant: string) => JSX.Element;
      const Container = (props: { tenant: string }) => viewVisualizationRenderer(props.tenant);
      const result = shallow(<Container tenant={'tenant1'} />);
      result.find('[data-test-subj="view-visualizations"]').simulate('click');
      expect(mockTenantUtils.selectTenant).toHaveBeenCalled();
    });

    it('render view dashboard column when tenant permissions contains multiple tenants or tenant pattern', () => {
      const tenantPermissions = [{ tenant_patterns: ['*'], permissionType: 'dummy' }];
      const wrapper = mount(
        <TenantsPanel
          roleName={sampleRoleName}
          tenantPermissions={tenantPermissions}
          errorFlag={false}
          coreStart={mockCoreStart as any}
          loading={false}
          isReserved={true}
        />
      );

      const prompt = wrapper
        .find('[data-test-subj="tenant-permission-container"]')
        .first()
        .getElement();
      const component = shallow(prompt);
      const columns = component.prop<
        Array<EuiTableFieldDataColumnType<RoleTenantPermissionDetail>>
      >('columns');
      const viewDashboardRenderer = columns[3].render as (tenant: string) => JSX.Element;
      const Container = (props: { tenant: string }) => viewDashboardRenderer(props.tenant);
      const result = shallow(<Container tenant={RoleViewTenantInvalidText} />);
      expect(result).toMatchSnapshot();
    });

    it('render view visualization column when tenant permissions contains multiple tenants or tenant pattern', () => {
      const tenantPermissions = [{ tenant_patterns: ['*'], permissionType: 'dummy' }];
      const wrapper = mount(
        <TenantsPanel
          roleName={sampleRoleName}
          tenantPermissions={tenantPermissions}
          errorFlag={false}
          coreStart={mockCoreStart as any}
          loading={false}
          isReserved={true}
        />
      );

      const prompt = wrapper
        .find('[data-test-subj="tenant-permission-container"]')
        .first()
        .getElement();
      const component = shallow(prompt);
      const columns = component.prop<
        Array<EuiTableFieldDataColumnType<RoleTenantPermissionDetail>>
      >('columns');
      const viewVisualizationRenderer = columns[4].render as (tenant: string) => JSX.Element;
      const Container = (props: { tenant: string }) => viewVisualizationRenderer(props.tenant);
      const result = shallow(<Container tenant={RoleViewTenantInvalidText} />);
      expect(result).toMatchSnapshot();
    });

    it('fetch data error', (done) => {
      jest.spyOn(React, 'useEffect').mockImplementationOnce((f) => f());
      // Hide the error message
      jest.spyOn(console, 'log').mockImplementationOnce(() => {});
      mockTenantUtils.fetchTenants.mockImplementationOnce(() => {
        throw Error();
      });

      shallow(
        <TenantsPanel
          roleName={sampleRoleName}
          tenantPermissions={[]}
          errorFlag={false}
          coreStart={mockCoreStart as any}
          loading={false}
          isReserved={false}
        />
      );

      process.nextTick(() => {
        // Expect setting error to true
        expect(setState).toHaveBeenCalledWith(true);
        done();
      });
    });

    it('fetch data', (done) => {
      jest.spyOn(React, 'useEffect').mockImplementationOnce((f) => f());

      const mockTenantPermissionData = [
        {
          tenant_patterns: ['tenant_1'],
          permissionType: 'read only',
          tenant: 'tenant_1',
          reserved: false,
          description: '',
          tenantValue: 'tenant_1',
        },
      ];

      mockTenantUtils.transformRoleTenantPermissionData = jest
        .fn()
        .mockReturnValue(mockTenantPermissionData);

      shallow(
        <TenantsPanel
          roleName={sampleRoleName}
          tenantPermissions={[]}
          errorFlag={false}
          coreStart={mockCoreStart as any}
          loading={false}
          isReserved={false}
        />
      );

      process.nextTick(() => {
        expect(mockTenantUtils.fetchTenants).toHaveBeenCalledTimes(1);
        expect(mockTenantUtils.transformTenantData).toHaveBeenCalledTimes(1);
        expect(mockTenantUtils.transformRoleTenantPermissionData).toHaveBeenCalledTimes(1);
        expect(mockAuthInfoUtils.getCurrentUser).toHaveBeenCalledTimes(1);
        expect(setState).toHaveBeenCalledWith(mockTenantPermissionData);
        done();
      });
    });
  });
});
