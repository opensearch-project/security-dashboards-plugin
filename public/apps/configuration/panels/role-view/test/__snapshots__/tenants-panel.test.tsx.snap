// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Role view - tenant panel Tenant permission list render view dashboard column when tenant permissions contains multiple tenants or tenant pattern 1`] = `
<Fragment>
  <EuiText
    size="s"
  >
    N/A
  </EuiText>
</Fragment>
`;

exports[`Role view - tenant panel Tenant permission list render view visualization column when tenant permissions contains multiple tenants or tenant pattern 1`] = `
<Fragment>
  <EuiText
    size="s"
  >
    N/A
  </EuiText>
</Fragment>
`;
